Flush any pending buffers on close()

ringerc opened the pull request:
The RemoteReporter buffers spans, but didn't flush them on close.
So any spans Finish()ed between the last flush interval and the
Close() of the Tracer would be lost.

Fixes #52

Signed-off-by: Craig Ringer <craig@2ndquadrant.com>

rnburn commented:
btw - is calling close automatically like this in the reporter's destructor really a good idea?
https://github.com/jaegertracing/cpp-client/blob/master/src/jaegertracing/reporters/RemoteReporter.h#L45
it can cause the process to take much longer to exit than it otherwise would.

ringerc commented:
On 5 February 2018 at 22:40, Ryan <notifications@github.com> wrote:

> btw - is calling close automatically like this in the reporter's
> destructor really a good idea?
> https://github.com/jaegertracing/cpp-client/blob/master/src/jaegertracing/
> reporters/RemoteReporter.h#L45
> it can cause the process to take much longer to exit than it otherwise
> would.
>

Yeah. Possibly not in the dtor; maybe it's best done only via an explicit
Close() from the Tracer. Need to expose internal API for that.

ringerc commented:
OK, so addressing #53 instead with an extension `Tracer.Flush()` method to allow explicit flushes.

ringerc commented:
Hrm, doesn't seem to do the job. I'm still losing spans. General approach seems sensible, forcing a blocking flush via separate API request.  I just haven't figured out how to do it yet through all the layers involved.

ringerc commented:
Maybe something like

```
 void RemoteReporter::flush()
{
    async_flush();
    std::unique_lock<std::mutex> lock(_mutex);
    _cv.wait(lock, [this]() {
        return !_running || _queue.empty();
    });
}
```

?

isaachier commented:
Actually, I see the issue is that I didn't add a "poison pill" to the end of the queue, instead used a boolean flag `running`. That means the shutdown is immediate. Will patch and submit new PR. Thanks for the help here, but will be taking a different approach, the one used in the [Go client](https://github.com/jaegertracing/jaeger-client-go/blob/master/reporter.go).

isaachier commented:
Actually, I see we do check for an empty queue in the thread's main loop, so I'm not sure why this isn't effectively the same thing. Also, now that I've read #53, I see this issue has further reaching goals than just fixing this local problem. Might be worth adding an eager flush mechanism as you suggested.

isaachier commented on src/jaegertracing/Tracer.h:
```diff
@@ -194,6 +194,11 @@ class Tracer : public opentracing::Tracer,
 
     void close() noexcept { Close(); }
 
+    void Flush()
```
I only use the title case functions for opentracing-cpp methods. I'd prefer `flush` here unless it is added to the OT implementation and this is an override.

ringerc replied:
Fixed in latest

ringerc commented:
I'm still seeing intermittent lost spans with this patch, though less frequently.

I just did some test runs with Wireshark observing the udp communications. The lost spans definitely never seem to get sent on the wire, they're not misplaced in the collector. The span name / operation name never appears on the wire at all.

Cause is uncertain as yet. Having trouble debugging effectively. May have to write small test program to see if I can repro.

ringerc commented:
OK, modified as discussed. It now flushes buffers on explicit `Close()` only.

My implementation of `RemoteReporter::flush()` was pretty much completely wrong. Fixed now, so it waits on its own condition variable and the reporter thread sets that condition variable after a wakeup (whether or not a span was flushed).

ringerc commented:
@isaachier Should be ready for review/merge

isaachier commented on src/jaegertracing/reporters/RemoteReporter.cpp:
```diff
@@ -137,5 +141,14 @@ void RemoteReporter::flush()
+{
+    std::unique_lock<std::mutex> lock(_mutex);
+    _cv_flush.wait(lock, [this]() {
+        async_flush();
```
This definitely looks wrong. The lambda here is just meant to express a boolean condition to avoid spurious wakeups. For example, instead of `while (running) { cv.wait(); }` you can do `cv.wait([&running] { return running; }`. This will potentially lead to many bugs in the program.

isaachier commented:
Fixed in #59.

ringerc commented:
Thanks!​
